### ISSUE TITLE
"Cannot read properties of null (reading 'postMessage')" while switching origin

### ISSUE THREAD

**Issue by xxluke** [cypress-io/cypress#28728]
### Current behavior

I get this error in about half of all attempts:

![grafik](https://github.com/cypress-io/cypress/assets/8194761/32116726-3faa-4fa1-b151-2428b53a7046)

In console:
> Uncaught (in promise) TypeError: Cannot read properties of null (reading 'postMessage')
>     at PrimaryOriginCommunicator.toSource (cypress_runner.js:169306:12)
>     at PrimaryOriginCommunicator.<anonymous> (index-f74a2fdb.js:146501:43)

While Cypress states that the "error originated from your application code, not from Cypress", it actually does originate from Cypress, here:
![grafik](https://github.com/cypress-io/cypress/assets/8194761/51df913b-b8be-45e2-bd7b-203de19aaff4)

### Desired behavior

There should be no error. Also it's misleading that the UI says the error doesn't originate from Cypress when it does.

### Test code to reproduce

I couldn't make a reproduction in time and I don't want to publish my company's internal url. However, the test code is very simple:
```ts
cy.visit('http://localhost:4200/'); // Used to set the global origin to localhost instead of the login provider, which would otherwise be the first visit. Not sure if relevant: localhost would js-redirect to company home page if given enough time.
cy.visit('https://sso-provider.example.com/login');
// Error
cy.origin('https://sso-provider.example.com', { args: loginData }, ({ user, password }) => {
  // ...
});
```
As you can see on my first screenshot, a XHR request of the first website is still marked as pending, so I guess the ending event was the one that couldn't get through anymore.

### Cypress Version

13.6.2

### Node version

v18.17.0

### Operating System

Windows 11 22H2

### Other

As a workaround, a global event listener _seems_ to work:
```ts
Cypress.on('uncaught:exception', (error: Error) => {
  // returning false here prevents Cypress from failing the test
  console.error('Caught error', error);
  if (error.stack?.includes('PrimaryOriginCommunicator.toSource')) {
    return false;
  }
  return true;
});
```

**Comment by msyniuk-reenbit** [comment]
facing same issue

**Comment by Sh1d0w** [comment]
Experiencing the same issue when trying to login in our app with Keycloak

**Comment by Elte156** [comment]
We are seeing the same issue as well. Some of our tests end with asserting that the page redirection (to another origin) worked. Then when the next test runs, we get this error intermittently.

`TypeError: Cannot read properties of null (reading 'postMessage')`

**Comment by dfakhim** [comment]
We're also have been experiencing same issues with 'postMessage' exception being thrown
<img width="947" alt="Screenshot 2024-04-04 at 11 18 29 AM" src="https://github.com/cypress-io/cypress/assets/65349580/2871348f-787d-4cb1-a0fa-28fdde0193e1">

**Comment by yulianovdey** [comment]
In our case listening for uncaught exceptions inside cy.origin stopped Cypress from running into the `postMessage` issue:

```javascript
cy.origin(Cypress.env('app_url'), () => {
  cy.on('uncaught:exception', (e) => {
    console.log(e);
    return false;
  });
});
```

**Comment by crice88** [comment]
So I had a similar issue but with our `cy.visit()` method. Up until recently we had had routes such as `cy.visit('logout')` or `cy.visit('login')`. After a minor upgrade to 13.4.0 (and an upgrade of Node 16 to 20), these methods began throwing the same errors that OP was receiving. Our solution was to just include leading forward slashes in our visit param i/e `cy.visit('/login')`

This doesn't solve OPs issue, but may help someone struggling with this error.

**Comment by AdeZwart** [comment]
We also have this issue. It's with only 1 very specific test. 

The weirdest thing is that it seems like it started randomly. No code changes on the repo, same versions of Node (18.19), Cypress(12.17) and Chrome (120) used.

**Comment by Jerryhu1** [comment]
> We also have this issue. It's with only 1 very specific test.
> 
> The weirdest thing is that it seems like it started randomly. No code changes on the repo, same versions of Node (18.19), Cypress(12.17) and Chrome (120) used.

Exact same issue here, Cypress 12.17, node 18.12. 1 specific test started failing suddenly without changes. 
Running the test in isolation passes without problems.

**Comment by tomasz-sobczyk-wttech** [comment]
We also have had this problem recently...

**Comment by mcmartincerny** [comment]
I ran into the same problem today. Do we know what does the "postMessage" even do? Is it important? Is the workaround mentioned viable?

**Comment by MKhasib** [comment]
@mcmartincerny My issue was clicking a link that opened a different origin(A different url than cypress). I fixed it by preventing such behavior 

I was using this:
cy.visit('/...')
So changing the origin will mean you can't visit the same docs for your app

**Comment by DancingCode** [comment]
I am seeing this issue on `cy.reload()`. Is there a fix please?

**Comment by Joshua-Hoban-visi-one** [comment]
I get this as well but I can't put a slash in front of my url as I use a full path, `cy.visit(https://....)`

**Comment by StefanVanDyck** [comment]
Not sure if this helps anyone else, but I had the exact same issue. 
Reading  https://www.cypress.io/blog/mistake-when-using-cy-session-and-how-to-solve-it solved it for me.

TLDR: 
Our custom login command was not waiting for the redirect from keycloak before storing the session.
Not sure if it was the session being corrupted somehow, or simply the act of switching during a redirect.
But providing a clean assertion after clicking login, to validate we followed the login redirect, completely fixed things.

**Comment by achandansoni** [comment]
any workaround for this issue I am having the same when one of my test redirects to another web page and then visits the same page again it throws the same error after the upgrade to latest version 14.2.7. and its random some times it passes.

**Comment by CSharpFiasco** [comment]
> any workaround for this issue I am having the same when one of my test redirects to another web page and then visits the same page again it throws the same error after the upgrade to latest version 14.2.7. and its random some times it passes.
>  

You can handle the exception 

cy.on('uncaught:exception', ...)

I'm on vacation this week so I can't get you the full snippet but there's cypress docs that'll help

**Comment by achandansoni** [comment]
> > any workaround for this issue I am having the same when one of my test redirects to another web page and then visits the same page again it throws the same error after the upgrade to latest version 14.2.7. and its random some times it passes.
> 
> You can handle the exception
> 
> cy.on('uncaught:exception', ...)
> 
> I'm on vacation this week so I can't get you the full snippet but there's cypress docs that'll help

I did the same workaround as the author and it worked thank you!
